File: Leetcode/0839 Similar String Groups.py
class Solution:
    def numSimilarGroups(self, strs):
      adj_list = build_graph(strs)
      n = len(strs)
      visited = [False] * n
      
      def dfs(start):
        if visited[start]:
          return
        visited[start] = True
        for neighbour in adj_list[start]:
          dfs(neighbour)
        
      n_groups = 0
      for node in range(n):
        if not visited[node]:
          dfs(node)
          n_groups += 1
      return n_groups

def build_graph(strs):
  n = len(strs)
  adj_list = [[] for _ in range(n)]
  for i in range(n):
    for j in range(i + 1, n):
      if is_similar(strs[i], strs[j]):
        adj_list[i].append(j)
        adj_list[j].append(i)
  return adj_list

def is_similar(a, b):
  diff = sum(map(lambda ab: ab[0] != ab[1], zip(a, b)))
  return diff == 0 or diff == 2

File: Leetcode/0839 Similar String Groups.ts
function numSimilarGroups(strs: string[]): number {
  const adjList = buildGraph(strs);
  const n = strs.length;
  const visited = new Array<boolean>(n).fill(false);
  function dfs(start: number) {
    if (visited[start]) {
      return;
    }
    visited[start] = true;
    for (const neighbour of adjList[start]) {
      dfs(neighbour);
    }
  }
  let nGroups = 0;
  for (let node = 0; node < n; node++) {
    if (!visited[node]) {
      dfs(node);
      nGroups++;
    }
  }
  return nGroups;
}

function buildGraph(strs: string[]) {
  const n = strs.length;
  const adjList = new Array<number[]>(n);
  for (let i = 0;i < n; i++) {
    adjList[i] = [];
  }
  for (let i = 0;i < n; i++) {
    for (let j = i + 1; j < n; j++) {
      if (isSimilar(strs[i], strs[j])) {
        adjList[i].push(j);
        adjList[j].push(i);
      }
    }
  }
  return adjList;
}

function isSimilar(a: string, b: string): boolean {
  let diff = 0;
  for (let i = 0; i < a.length; i++) {
    if (a[i] !== b[i]) {
      diff++;
    }
  }
  return diff === 0 || diff == 2;
}
